File: Sort/Quick_Sort/quickSort_test.java
package Sort.Quick_Sort;

import java.util.Scanner;

public class quickSort_test {
    public void quickSort(int[] arr, int start, int end) {
        if (start >= end) return;
        int mid = partition_v1(arr, start, end);

        quickSort(arr, start, mid);
        quickSort(arr, mid+1, end);
    }

    private int partition_v1(int[] arr, int start, int end) {
        if (start >= end) return start;
        int pivot = arr[start];
        int i = start+1, j = end;
        while (i <= j) {
            while (i <= j && arr[i] < pivot) i++;
            while (i <= j && arr[j] > pivot) j--;
            if (i < j) {
                swap(arr, i, j);
            }
        }
        swap(arr, start, j);
        return j;
    }

    private void swap (int[] arr, int i, int j) {
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
            // System.out.println(arr[i]);
        }

        quickSort_test sol = new quickSort_test();
        sol.quickSort(arr, 0, n-1);
        for (int num: arr)
            System.out.print(num+" ");
    }
}

File: Sort/Quick_Sort/quickSort_test.c
#include <stdio.h>
#include <stdlib.h>
// #include <string.h>
// #include <stdbool.h>
// #include <math.h>
// #include <limits.h>

// quickSort_test.c, 类比q75

void swap(int* a, int* b);

// 快排v1: 显式partition()返回mid下标（右闭）
int partition(int* nums, int start, int end);

void quickSort_v1(int* nums, int start, int end) {
    if (start >= end) return;
    int mid = partition(nums, start, end);
    quickSort_v1(nums, start, mid); // 右闭
    quickSort_v1(nums, mid + 1, end);
}

int partition(int* nums, int start, int end) {
    int pivot = nums[start];
    int i = start, j = end;
    while (i <= j) {
        while (i <= j && nums[i] < pivot) i++;
        while (i <= j && nums[j] > pivot) j--; //?严格不等于pivot! 否则，有重复元素将陷入死循环!
        if (i <= j) { // ?严格i<=j
            swap(&nums[i], &nums[j]);
            i++; j--;// ?勿忘i++; j--; 否则，会陷入死循环!
        }
    }
    swap(&nums[start], &nums[j]);
    return j; // 或i-1, 即分区mid下标（右闭）
}

void swap(int* a, int* b) {
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

// 快排v2: 隐式partition()
void quickSort_v2(int* nums, int start, int end) {
    if (start >= end) return;
    if (start >= end) return;
    int i = start, j = end;
    int pivot = nums[(start + end) / 2]; // pivot取中间，避免最后swap([start], [j])
    while (i <= j) {
        while (i <= j && nums[i] < pivot) i++;
        while (i <= j && nums[j] > pivot) j--;
        if (i <= j) {
            swap(&nums[i], &nums[j]);
            i++; j--;
        }
    }
    // swap(&nums[start], &nums[j]); // 当pivot取nums[start]时, 需要
    quickSort_v2(nums, start, j); // 隐式partition[start, R], [L/R+1, end]
    quickSort_v2(nums, i, end);
}



int main() {
    int nums[] = { 2,3,2,1,5,0 };
    int n = sizeof(nums) / sizeof(*nums);
    quickSort_v1(nums, 0, n - 1);
    for (int i = 0; i < n; i++) {
        printf("%d ", nums[i]);
    }
    printf("\n");

    int nums2[] = { 2,3,2,1,5,0 };
    quickSort_v2(nums2, 0, n - 1);
    for (int i = 0; i < n; i++) {
        printf("%d ", nums2[i]);
    }
    printf("\n");
    return 0;
}